snuba/subscriptions/scheduler.py

```diff
@@ -39,3 +42,47 @@ def find(self, interval: Interval[datetime]) -> Iterator[ScheduledTask[TTask]]:
         ascending order.
         """
         raise NotImplementedError
+
+
+# XXX: This is all temporary until we get these types properly introduced
+PartitionId = NewType("PartitionId", int)
+
+
+@dataclass(frozen=True)
+class SubscriptionIdentifier:
+    partition: PartitionId
+    uuid: UUID
+
+
+class Subscription(NamedTuple):
+    identifier: SubscriptionIdentifier
+    data: SubscriptionData
+
+
+class SubscriptionScheduler(Scheduler[Subscription]):
+    def __init__(self) -> None:
+        self.__subscriptions: Dict[UUID, Subscription] = {}
+
+    def set(self, value: Subscription) -> None:
+        self.__subscriptions[value.identifier.uuid] = value
+
+    def delete(self, key: UUID) -> None:
+        self.__subscriptions.pop(key, None)
+
+    def clear(self) -> None:
+        # XXX: Not part of the standard interface. Since we want to periodically clear
+        # the cache, whatever is managing this should probably clear and refetch
+        self.__subscriptions = {}
```

Comment:
Should these be internal methods? I kind of assumed the constructor would just be passed the `RedisSubscriptionStore` and any sort of caching of that state would happen privately in the implementation, curious if you have a different approach in mind.

Reply:
I based these extra methods off of https://github.com/getsentry/snuba/compare/scheduler#diff-22926dd5eda88654f5e7bb17ae6165aaR8-R26, and assumed that something external would be passing updates in (like the `SubscriptionConsumer` or whatever we end up calling it).

I don't mind too much if this class just gets passed a datastore and receives updates from it though.

Reply:
Hmm, passing in the datastore doesn't work so well after your refactor, since we expect this to return `Iterator[ScheduledTask[Subscription]]`, and the datastore can only really return `SubscriptionData` rows. I suppose that if we pass the partition id to this class then we can rebuild the `SubscriptionIdentifier` with that + the UUID. Thoughts?

Reply:
> I suppose that if we pass the partition id to this class then we can rebuild the SubscriptionIdentifier with that + the UUID. Thoughts?

Yeah, I think that is fine. I don't love that both the store and the scheduler will have the partition ID as attributes, but that should be able to be changed reasonably easily later if it's a problem. The only other options that I can see would be to either:

1. expose the scheduler interface on `RedisSubscriptionDataStore`, or
2. make the subscription store deal with `SubscriptionIdentifier` keys instead of `UUID` ones (this one doesn't seem so bad, to be honest)

Reply:
2. Doesn't seem so bad... Slight redundancy in the keys, but that's all. Kind of weird though - if we did that, then we may as well just pass the `Subscription` object when creating?

Reply:
Gonna just go with passing the partition id for now. We can change it as needed (and after your other pr lands)

Reply:
Yeah, I think ideally the input data structure matches the output one, so `Subscription` objects would go in, and `Subscription` objects would come out.

That would make it so there would be a single interface for dealing with all of the partitions at once on the write path, which is kind of nice — we could just set it in once in the constructor of the `Creator` and not have to instantiate it all of the time.

It is a bit problematic for the worker, which only needs the subscriptions from a single partition when scheduling ([example here](https://github.com/getsentry/snuba/blob/388a89cd7507bd9667425229fbc42859e42ab5c2/snuba/subscriptions/worker.py#L36-L45), this commit may be GC'ed in the future due to rebasing). Ideally we wouldn't have to filter out results from `store.all` or `scheduler.find` based on their partition, so we might need to put a bit more consideration into how these interact.